### ISSUE TITLE
Dynamic Schema Selection For oneOf Error Message Reduction

### ISSUE THREAD

**Issue by nmstone** [Eomm/fastify-schema-constraint#35]
Hello! I have been investigating this plugin and am wondering if it is the right fit.

Here is my use case:

I would like to have a single fastify route with a schema that validates against many different inputs. I have this working already with a oneOf in the body of my JSON schema, however the validation error messages I get are not helpful, as I end up with many messages describing why each oneOf entry failed.

I was hoping that this plugin would allow me to dynamically select a schema, that way I would only get validation errors for the matching schema, however the oneOf used by the plugin seems to cause the exact same issue regarding error messages. I guess I have basically only moved the oneOf up a level - from deeply nested in my body schema, to select from multiple different bodies?

As an example, this is the kind of error that I am looking to avoid and is produced by the code below: `body should have required property 'token', body should have required property 'topic', body should match exactly one schema in oneOf`

In the above error, the schema would ideally be dynamically selected so that I would only get the relevant error from the selected schema in the oneOf, for example just: `body should have required property 'token'`

Is what I am looking for (the reduction of irrelevant errors from validation) possible using this plugin?

Here is some code using this plugin for context:

```
const constraint = {
  body: {
    constraint: function (request) {
      switch(request.params.type){
        case 'schema1': return '#schema1'
        case 'schema2': return '#schema2'
        default: return null
      }
    },
  },
}

... register ...

const routeSchema = {
  oneOf: [
    { $id: '#schema1', type: 'object', properties: { token: { type: 'string' } }, required: ['token'] },
    { $id: '#schema2', type: 'object', properties: { topic: { type: 'string' } }, required: ['topic'] },
  ]
}

// I have a factory that generates a valid fastify server 
// from routes defined as objects like the one below - hence the
// different formatting to a usual fastify route

const test = {
  url: '/test/:type',
  method: 'POST',
  handler: (request, reply) => {
    reply.send('hi')
  },
  schema: {
    body: routeSchema,
    params: {
      type: 'object',
      properties: {
        type: {
          type: 'string',
          enum: ['schema1', 'schema2'],
        }
      },
      required: ['type'],
    }
  },
}
```

**Comment by nmstone** [comment]
TLDR; Solved by reworking my fastify route's JSON Schema to handle conditional subschemas based on a property value rather than adding conditional schema support to a single fastify route through this plugin.

I ended up not needing to use this plugin and using the `anyOf: [ if/then ... if/then]` approach in a single deeply nested conditional subschema JSON Schema as described here: https://json-schema.org/understanding-json-schema/reference/conditionals.html#if-then-else

This worked to summarize my error messages to just the error message from the matching conditional subschema (based on a property value defined in the if) rather than enumerating errors for why each of the subschemas in the oneOf failed. Not sure if that makes any sense but maybe it will help someone going down the same path as me!

If I have some time i'll look into if the constraint oneOf in this plugin can be replaced with the above anyOf if/then approach.

**Comment by Eomm** [comment]
Happy you have solved.

I thought it was doable actually since one single schema is executed as validation so, you should not see any extra error from others `oneOf` options.

> If I have some time i'll look into if the constraint oneOf in this plugin can be replaced with the above anyOf if/then approach.

What the if/then approach lacks is the possibility to read the request's data.
Here the concept is to not try all the schemas for every request, but just the right one.
